content/blog/managing-google-workspace-with-terraform.md

```diff
@@ -0,0 +1,295 @@
+    "https://www.googleapis.com/auth/admin.directory.userschema",
+    "https://www.googleapis.com/auth/apps.groups.settings",
+    "https://www.googleapis.com/auth/iam",
+  ]
+}
+
+provider "googleworkspace" {
+  customer_id             = "my_customer"
+  credentials             = local.secrets["googleworkspace_admin_credentials_json"]
+  impersonated_user_email = var.impersonated_user_email
+  oauth_scopes            = local.oauth_scopes
+}
+
+# Get users and groups from YAML files. You might choose to use other config file.
+# Note: path.module refers to the directory where terraform/tofu plan/apply is run
+# var.googleworkspace_configs is a variable for the directory containing config files
+locals {
+  config_path = "${path.module}/${var.googleworkspace_configs}"
+  all_groups = yamldecode(file("${local.config_path}/groups.yaml"))
+  all_users  = yamldecode(file("${local.config_path}/users.yaml"))
+
+  # Skip objects that start with "_", which we use as default or prototype objects
+  groups = { for k, v in local.all_groups : k => v if !startswith(k, "_") }
+  users  = { for k, v in local.all_users : k => v if !startswith(k, "_") }
+}
+
+module "googleworkspace_users_groups" {
+  source  = "masterpointio/users-groups-automation/googleworkspace"
+  version = "X.X.X"
+  groups  = local.groups
+  users   = local.users
+}
+```
+
+## Module Design Decisions
+
+Below we point out a few Terraform module design decisions we made to reduce friction as others use the module.
+
+### Design Decision #1 - Implementing Integration and Validation Tests
+
+To ensure the Terraform module remains reliable after changes—whether by contributors or automated processes—we've added approximately 20 tests. Thankfully [Terraform](https://developer.hashicorp.com/terraform/tutorials/configuration-language/test#prerequisites) and [OpenTofu](https://opentofu.org/docs/cli/commands/test/) both include a built-in testing framework that allows us to write tests directly in HCL, replacing the previous Go-based approach. The tests validate that the module behaves as expected and prevent new changes from unintentionally breaking functionality that downstream consumers rely on.
+
+We'll skip over the happy path tests ([example 1](https://github.com/masterpointio/terraform-googleworkspace-users-groups-automation/blob/9b9112a2d3ff0f3d415813a8b4de11cefcab56a9/tests/variables_users.tftest.hcl#L52-L70), [example 2](https://github.com/masterpointio/terraform-googleworkspace-users-groups-automation/blob/9b9112a2d3ff0f3d415813a8b4de11cefcab56a9/tests/variables_users.tftest.hcl#L72-L89), [example 3](https://github.com/masterpointio/terraform-googleworkspace-users-groups-automation/blob/9b9112a2d3ff0f3d415813a8b4de11cefcab56a9/tests/variables_users.tftest.hcl#L140-L157)) we added to validate things like email and password validation, since these basic unit tests can be easily generated using Cursor, GitHub Copilot, or other AI code generation tools.
```

Comment:
is it intentional to point to a specific commit here?

Reply:
Yes. I imagine we'll change the tests in the future and I wanted the content to still make sense after code changes to the tests file.